doc/lsst.ip.diffim/code_notes.rst

```diff
@@ -0,0 +1,183 @@
+ignored. The following example specifies one science exposure and
+implicitly the deep coadd residing in the same repository for image
+differencing:
+
+.. code-block:: python
+
+	imageDifference.py repo/ingested/ --id visit=410915 filter=g
+	ccd=25
+
+``GetCoaddAsTemplate`` also supports loading ``DcrModel`` coadds and
+produce a color corrected template matching the ``filter`` of the
+science exposure. 
+	
+Specifying a calexp as template
+-------------------------------
+
+Using the ``GetCalexpAsTemplate`` subtask, we can select two calexp
+data products for subtraction.  The ``--id`` option specifies a
+science exposure.  The ``--templateId`` specifies a new *visit id*
+only. All additional field values are silently ignored; instead, the
+template calexp specification inherits its other field values from the
+``--id`` option.  The following example subtract visit 410915 ccd 25
+(template) from 411055 ccd 25 (science exposure):
+  
+.. code-block:: python
+
+	imageDifference.py repo/calibimgs --id visit=411055
+		ccd=25 --templateId visit=410915
+
+Implementation notes
+====================
+		
+General
+-------
+
+If the two images have the same origin and extent (same WCS) the
+subtraction is performed pixel by pixel. Otherwise, the *template*
+exposure is warped and resampled into the WCS frame of the science
+exposure.
+
+The AL kernel fitting is entirely implemented in C++, an adaptation of
+the `hotpants <https://github.com/acbecker/hotpants>`_ C package
+by Becker, while the ZOGY algorithm is entirely implemented in
+Pyton based on mostly on the ``numpy`` FFT functionality.
+
+Beside the top level configuration options, the AL algorithm has its
+own set of separate configuration parameters, while the ZOGY algorithm
+does not have any algorithm specific options.
+
+Alard-Lupton algorithm 
+----------------------
+
+The Alard-Lupton numerical fitting for the coefficients (of the
+composing basis kernel functions) of the image convolution kernel is
+performed independently on a number of stamp images around selected
+sources. These independent solutions are called *kernel candidates*
+throughout the code. The final kernel solution model for the whole
+image is a smoothly varying spatial fit of the coefficients as a
+function of the image pixel coordinates.
+
+If ``convolveTemplate==False`` the science exposure is convolved and
+then the template is subtracted from the science image.
+
+The performance of the AL algorithm was studied in details in the
+[Becker_LDM-227]_ report. This study forms the basis of the AL
+algorithm default values; the the degree of the polynomial
+multiplicator of the Gaussian kernel basis functions (``degGauss``),
+the degree of the polynomial that is fitted to the spatial variaton of
+the solution coefficients accross the image (``spatialKernelOrder``)
+and the default detection thresholds (5.5 sigma). As a legacy of this
+study, the ``Winter2013ImageDifferenceTask`` is still available in
```

Comment:
Yikes... 😨

Reply:
Lots of refactoring could be done to ip_diffim...